File: src/DesignPattern/Creational/AbstractFactory/AbstractFactory-java/src/widget/Widget.java
package widget;

interface Button {
    void drawButton();
}

interface Menu {
    void drawMenu();
}

// -------------------------------------------

class LinuxButton implements Button {
    public void drawButton() {
        System.out.println("draw LinuxButton");
    }
}

class WindowsButton implements Button {
    public void drawButton() {
        System.out.println("draw WindowsButton");
    }
}

class LinuxMenu implements Menu {
    public void drawMenu() {
        System.out.println("draw LinuxMenu");
    }
}

class WindowsMenu implements Menu {
    public void drawMenu() {
        System.out.println("draw WindowsMenu");
    }
}

// -------------------------------------

interface Factory {
    Button createButton();

    Menu createMenu();
}

class LinuxFactory implements Factory {
    public Button createButton() {
        return new LinuxButton();
    }

    public Menu createMenu() {
        return new LinuxMenu();
    }
}

class WindowsFactory implements Factory {
    public Button createButton() {
        return new WindowsButton();
    }

    public Menu createMenu() {
        return new WindowsMenu();
    }
}

// ------------------------------------

public class Widget {

    public static enum type {
        linux, windows,
    }

    private Button button;
    private Menu menu;

    public Widget(type t) {
        Factory factory = null;
        switch (t) {
        case linux:
            factory = new LinuxFactory();
            break;
        case windows:
            factory = new WindowsFactory();
            break;
        }
        this.button = factory.createButton();
        this.menu = factory.createMenu();
    }

    public void draw() {
        button.drawButton();
        menu.drawMenu();
    }
}

File: src/DesignPattern/Creational/AbstractFactory/AbstractFactory-go/src/widget/widget.go
package widget

import "fmt"

type Button interface {
	DrawButton()
}

type Menu interface {
	DrawMenu()
}

//---------------------------------------------
type linuxButton struct{}

func (p *linuxButton) DrawButton() {
	fmt.Println("draw LinuxButton")
}

type linuxMenu struct{}

func (p *linuxMenu) DrawMenu() {
	fmt.Println("draw linuxMenu")
}

type windowsButton struct{}

func (p *windowsButton) DrawButton() {
	fmt.Println("draw windowsButton")
}

type windowsMenu struct{}

func (p *windowsMenu) DrawMenu() {
	fmt.Println("draw windowsMenu")
}

type Factory interface {
	CreateButton() Button
	CreateMenu() Menu
}

type LinuxFactory struct{}

func (p *LinuxFactory) CreateButton() Button {
	return &linuxButton{}
}

func (p *LinuxFactory) CreateMenu() Menu {
	return &linuxMenu{}
}

type WindowsFactory struct{}

func (p *WindowsFactory) CreateButton() Button {
	return &windowsButton{}
}

func (p *WindowsFactory) CreateMenu() Menu {
	return &windowsMenu{}
}
